registry/storage/driver/s3-aws/s3.go

```diff
@@ -942,15 +942,152 @@ func (d *driver) Walk(ctx context.Context, from string, f storagedriver.WalkFn)
+}
+
+// ModTime returns the modification time for the file. For backends that
+// don't have a modification time, the creation time should be returned.
+func (wi walkInfoContainer) ModTime() time.Time {
+	return wi.FileInfoFields.ModTime
+}
+
+// IsDir returns true if the path is a directory.
+func (wi walkInfoContainer) IsDir() bool {
+	return wi.FileInfoFields.IsDir
+}
+
+// doWalkRecursive is an implementation of Walk that is optimized for listing high level directories, such as repositories,
+// where storagedriver.ErrSkipDir is utilized to skip over large numbers of keys.
+func (d *driver) doWalkRecursive(parentCtx context.Context, objectCount *int64, path, prefix string, f storagedriver.WalkFn) (bool, error) {
+	var retError error
+	var retStop bool
+
+	listObjectsInput := &s3.ListObjectsV2Input{
+		Bucket:    aws.String(d.Bucket),
+		Prefix:    aws.String(path),
+		Delimiter: aws.String("/"),
+		MaxKeys:   aws.Int64(listMax),
+	}
+
+	ctx, done := dcontext.WithTrace(parentCtx)
+	defer done("s3aws.ListObjectsV2Pages(%s)", path)
+	listObjectErr := d.S3.ListObjectsV2PagesWithContext(ctx, listObjectsInput, func(objects *s3.ListObjectsV2Output, lastPage bool) bool {
+		// KeyCount was introduced with version 2 of the GET Bucket operation in S3.
+		// Some s3 implementations (looking at you ceph/rgw) have a buggy
+		// implementation so we intionally avoid ever using it, preferring instead
+		// to calculate the count from the Contents and CommonPrefixes fields of
+		// the s3.ListObjectsV2Output. we retain the commented out KeyCount code
+		// and this comment so as not to forget this problem moving forward.
+		//
+		// count = *objects.KeyCount
+		// *objectCount += *objects.KeyCount
+		count := int64(len(objects.Contents) + len(objects.CommonPrefixes))
+		*objectCount += count
+
+		walkInfos := make([]walkInfoContainer, 0, count)
+
+		for _, dir := range objects.CommonPrefixes {
+			commonPrefix := *dir.Prefix
+			walkInfos = append(walkInfos, walkInfoContainer{
+				prefix: dir.Prefix,
+				FileInfoFields: storagedriver.FileInfoFields{
+					IsDir: true,
+					Path:  strings.Replace(commonPrefix[:len(commonPrefix)-1], d.s3Path(""), prefix, 1),
+				},
+			})
+		}
+
+		for _, file := range objects.Contents {
+			// empty prefixes are listed as objects inside its own prefix.
+			// https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html
+			if strings.HasSuffix(*file.Key, "/") {
+				continue
+			}
+			walkInfos = append(walkInfos, walkInfoContainer{
+				FileInfoFields: storagedriver.FileInfoFields{
+					IsDir:   false,
+					Size:    *file.Size,
+					ModTime: *file.LastModified,
+					Path:    strings.Replace(*file.Key, d.s3Path(""), prefix, 1),
+				},
+			})
+		}
+
+		sort.SliceStable(walkInfos, func(i, j int) bool { return walkInfos[i].FileInfoFields.Path < walkInfos[j].FileInfoFields.Path })
+
+		for _, walkInfo := range walkInfos {
+			err := f(walkInfo)
+
+			if err == storagedriver.ErrSkipDir {
```

Comment:
Can we make this `error` handling more like idiomatic `Go` (whatever that means) i.e.
```Go
if err != nil {
  if err == storagedriver.ErrSkipDir {
    // do things
  }
  // do other things
}
```